src/VisualStudio/LiveShare/Impl/Client/RemoteLanguageServiceWorkspace.cs

```diff
@@ -94,100 +113,134 @@ void IRunningDocumentTableEventListener.OnRenameDocument(string newMoniker, stri
+
             session.RemoteServicesChanged += (object sender, RemoteServicesChangedEventArgs e) =>
             {
                 _remoteDiagnosticListTable.UpdateWorkspaceDiagnosticsPresent(_session.RemoteServiceNames.Contains("workspaceDiagnostics"));
             };
         }
 
-        public void EndSession()
-        {
-            IsRemoteSession = false;
-        }
-
-        public void Init()
-        {
-            StartSolutionCrawler();
-        }
+        public string? GetRemoteExternalRoot(string filePath)
+            => _registeredExternalPaths.SingleOrDefault(externalPath => filePath.StartsWith(externalPath));
 
-        /// <inheritdoc />
-        public override bool CanOpenDocuments => true;
+        public string? GetRemoteWorkspaceRoot(string filePath)
+            => _remoteWorkspaceRootPaths.SingleOrDefault(remoteWorkspaceRoot => filePath.StartsWith(remoteWorkspaceRoot));
 
-        /// <inheritdoc />
-        public void OnManagedProjectAdded(ProjectInfo projectInfo)
+        /// <summary>
+        /// Event that gets triggered whenever the active workspace changes.  If we're in a live share session
+        /// this means that the remote workpace roots have also changed and need to be updated.
+        /// This will not be called concurrently.
+        /// </summary>
+        private async Task OnActiveWorkspaceChangedAsync(object sender, EventArgs args)
         {
-            //Notify AnyCode Roslyn workspace a project is being added.
-            OnProjectAdded(projectInfo);
+            if (IsRemoteSession)
+            {
+                await UpdatePathsToRemoteFilesAsync(_session!).ConfigureAwait(false);
+            }
         }
 
-        /// <inheritdoc />
-        public void OnManagedProjectReloaded(ProjectInfo projectInfo)
+        /// <summary>
+        /// Retrieves the base folder paths for files on the client that have been retrieved from the remote host.
+        /// Triggers a refresh of all open files so we make sure they are in the correct workspace.
+        /// </summary>
+        private async Task UpdatePathsToRemoteFilesAsync(CollaborationSession session)
         {
-            //Notify AnyCode Roslyn workspace a project is being reloaded.
-            OnProjectReloaded(projectInfo);
+            var (remoteRootPaths, externalPaths) = await GetLocalPathsOfRemoteRoots(session).ConfigureAwait(false);
+
+            // Make sure we update our references to the remote roots and iterate RDT only one at a time.
+            using (await s_RemotePathsGate.DisposableWaitAsync(CancellationToken.None).ConfigureAwait(false))
+            {
+                if (IsRemoteSession && (!_remoteWorkspaceRootPaths.Equals(remoteRootPaths) || !_registeredExternalPaths.Equals(externalPaths)))
+                {
+                    _remoteWorkspaceRootPaths = remoteRootPaths;
+                    _registeredExternalPaths = externalPaths;
+                    await RefreshAllFilesAsync().ConfigureAwait(false);
```

Comment:
This is _potentially_ a deadlock risk: if the UI thread is unavailable to switch to, we're still holding the lock, and so further notifications will get blocked up. Whether this is a serious concern is hard to say -- if we're wrong we'll at least get a nice hang which should be easy to debug. :smile: